Feature/launch template metadata

bird-bust-battery opened the pull request:
Hello! I would like to add the ability to specify metadata options for the instances that are created through this module.

Backwards-compatibility considerations:
- `var.bastion_metadata_options` is optional and defaults to no keys.
- The `metadata_options` block on the `aws_launch_template` resource will only be added if one or more `var.bastion_metadata_options` keys are provided.
- Defaults for `http_endpoint`, `http_tokens`, and `http_put_response_hop_limit` match the current version of the AWS provider ([v5.66.0](https://registry.terraform.io/providers/hashicorp/aws/5.66.0/docs/resources/launch_template#metadata-options)).
- Support for the last of the five specified `metadata_options` keys was added in provider version v3.72.0, which is older than the version mentioned in the README.

This was tested using Terraform version 1.9.4. I validated against 1.2.0 but have not deployed with the older version.

bird-bust-battery committed:
Add support for launch template metadata options.

bird-bust-battery committed:
Update README.

joshuamkite commented:
Thanks for the Pull request!

I think adding the `metadata_options block`  as an optional feature is a good idea but I am not sure about this implementation. I'm not using this module myself in production right now so please forgive me if I have misunderstood anything about this component.  

IIRC all of the values in the block  are described as 'optional' in the documentation and therefore it should be possible to specify an arbitrary subset of them. Currently in this P/R the block itself is optional and not included by default, which is great, but the values for this block are defaulted. This means that the complete set will always be applied, with this module's default values where not specified. Whilst some of these do have an API default value where not specified, not all of them appear to. Ideally I prefer to avoid recreating default behaviour where possible in favour of just using that default behaviour instead. Would it be better if instead of using 

`type = map(any)` and `lookup`

We used 

`type = map(object({` with `optional(string)` and `try(each.value.${thing}, null)`? 

You can see examples of this approach [here](https://github.com/joshuamkite/terraform-aws-iam-role)

Alternatively, since there is no input validation being applied, I suppose it would be possible as a fallback to reference the block to the variable directly, without lookups, and have the variable default as simply an empty `map(any)` with the description advising the user of the options. This would similarly allow all the options you suggest providing, but leveraging the default behaviours rather than recreating them.

In either case I am not clear why we need `dynamic` block for a static set of attributes. 

What do you think?

Separately, things I will need to add to 'contributing' for future but shouldn't affect this P/R: 

- I will generally test with Open Tofu current version as a starting point
- `examples` will need module version incrementing with new version and deploying to check validity - I will expect to handle this but would welcome suggestions with P/Rs!
- If the feature change is significant then it may be worth considering adding another example deployment to test it (I don't think this is needed here)

Thanks!

bird-bust-battery committed:
Remove dynamic block. Do not supply any defaults.

bird-bust-battery commented:
Thanks for the feedback! I've pushed up a couple of changes.
- I'd also prefer using `object` with `optional` keys, but `optional` wasn't officially added to Terraform until 1.3. To make it compatible with 1.2, I'd need to set the noisy `experiments` key in the main `terraform` block.
- Sounds good with not assuming defaults. I've removed them in favor of `null`, and swapped out `lookup` with `try`.
- The dynamic block was an attempt to keep the empty `+ metadata_options {}` block from showing up in plans and applies for anyone not using this feature. I've removed it and am now just using a normal `metadata_options` block.

joshuamkite commented:
Thanks for this. I don't see a specific need to limit to an arbitrary historic version of Terraform- If you are testing with modern/current version of terraform and I am testing with modern/current version of OpenTofu then it should be fine.  People can always reference a previous version of the module. Does any of this change your suggestions? I had not thought through  your previous use of `dynamic` so perhaps that is prudent after all.

bird-bust-battery committed:
Bump min TF to 1.3. Update input variable. Add dynamic block. Update changelog.

bird-bust-battery commented:
That works for me. Overview of latest push:
- The `dynamic` block has been re-added with a comment to clarify its use.
- The input variable is now an `object`, with `optional` keys.
- The minimum Terraform version has been bumped to 1.3.0 to avoid issues with anyone still on 1.2.
- Changelog updated, with `TBD` added as the release version.

joshuamkite commented:
**Updated**

Thanks for these updates. I have been doing some testing with this branch (using `example/full_with_public_ip`) ~~and the result is, to me, a bit non obvious~~.

## Case 1 - Historic Default

The logic to ignore `metadata_options` if not declared works, so historic behaviour is not impacted, great!

## Case 2 - Partial configuration**s**, preserving the API defaults

This is where things seem to get a bit tricky, compounded by the fact that deploying updates in place to this block don't reliably propagate to new Autoscaling Group Instances. To address this I have been deploying with a configuration, attempt to connect, debug etc, and then teardown before deploying with a new configuration.

This works:

```terraform
  bastion_metadata_options = {
    http_endpoint               = "enabled"
    http_tokens                 = "optional"
    http_put_response_hop_limit = 1
    instance_metadata_tags      = "enabled"
  }
```

And I believe so does this:

```terraform
  bastion_metadata_options = {
    http_endpoint               = "enabled"
    http_tokens                 = "optional"
    http_put_response_hop_limit = 1
  }
```

This shows that the `dynamic` block with `try`/`optional` is working - also great!

**Updated**

This configuration also works:

```terraform
  bastion_metadata_options = {
    instance_metadata_tags = "enabled"
  }
```

and with it after deployment `terraform show` shows:

```terraform
   metadata_options {
        http_put_response_hop_limit = 0
        instance_metadata_tags      = "enabled"
    }
```

This is great, also - shows the API defaults working or themselves to a successful configuration.

~~However~~ **I note that** toggling the value of either of the first two attributes- `http_endpoint` ; `http_tokens` - breaks the service. Using `ssh -v` it's apparent that the connection is terminating at the host rather than the container (host agent version includes `Debian`). I haven't at this point got into looking at the logs on the host but it would seem that the container build is either not starting or not completing. ~~I would like to address this before release.~~ I am not current with the impact of these options in depth and so would appreciate your insight/additional input here. It would seem that one of the following is likely true:

1. Availability of the metadata service is a requirement for the bastion service to deploy and run successfully ~~- in which case I feel we should~~ . **By** default **the `metadata_options`** block **is set** to a working configuration when applied, which is great. Happy that people should understand the implication of exotic options if they want to apply them.
2. Setting either `http_endpoint = "disabled"` and/or `http_tokens = "required"` either always break deployment **OR** requires something else to _also_ be configured in order to enable a working service deployment. ~~In such case~~ I would propose ~~defaulting to a working configuration as per (1) above but accompanied with~~ some documentation advising of this.

~~Apologies if this seems pedantic, I do think that this is a useful feature to have but also that it can be a footgun for the unwary!~~

joshuamkite commented:
Apologies for earlier misfire. I have revisted this and the code seems fine, thanks! I have updated my previous comments in place with strikethrough and bold. I think it would be good to document that putting:

```hcl
    http_endpoint = "disabled"
    http_tokens   = "required"
```

without other changes breaks the deployment - perhaps something in the variable `description`?

bird-bust-battery commented:
I appreciate the deep dive on this.

Assuming the metadata service is required, would it make sense to:
- Remove the ability to set `http_endpoint` to `disabled`.
- Validate that `http_put_response_hop_limit` is within the valid range (1-64).

I think that when setting `http_tokens` by itself, `http_put_response_hop_limit` is being set to 0. I'm guessing that essentially disables the service since all hops exceed the limit.

Either way, I'll expand on the variable description to document where we land on this.

joshuamkite commented:
Thanks for this. I am not sure if or why the metadata service should be required myself. Personally, just as I would prefer to avoid duplicating default API behaviour in favour of using the default, I would generaly prefer to avoid disabling behaviour simply because I haven't considered potential use cases for it. 

I would question if setting `hops=0` does what you think since as you can see from my earlier comment, this value can be returned with `terraform show` for a working configuration even when explicitly set.

I presume that you were motivated to add this feature because you have some use case for it. Are you able to expand on what this might be and how you are using it as this might inform us better how to proceed?

bird-bust-battery commented:
> I presume that you were motivated to add this feature because you have some use case for it. Are you able to expand on what this might be and how you are using it as this might inform us better how to proceed?

Sure. One of our teams are using AWS resources where our usual method of secure access isn't supported, and this module is a potential alternative. We have to comply with internal security standards to use it, and one of those is requiring IMDSv2 on EC2 instances.

I'm moving onto some other work, so I don't know when I'll be able to spend some time digging into the metadata service details. Feel free to close if needed, and I'll update or reopen when I have some more time to look into this. Thanks for your help.

joshuamkite commented:
OK, TBH I think we are almost there, I can merge this to a branch and add the documentation there. Thanks for the contribution
